[MRG] Add label accuracy to micro-average in classification report

eamanu opened the pull request:
Add label accuracy to micro-average


#### Reference Issues/PRs
Fixes #12334

#### What does this implement/fix? Explain your changes.


#### Any other comments?
None

eamanu committed:
Add label accuracy to micro-average in classification report Issue:#12334

eamanu committed:
fix comment

jnothman commented on sklearn/metrics/classification.py:
```diff
@@ -1530,6 +1530,12 @@ class 2       1.00      0.67      0.80         3
+    # labelled micro average
+    label_micro_average = False
+    if y_type.startswith('multiclass') and \
+       (not labels_given or (len(labels) == len(target_names))):
```
I think you need more like `set(labels) == set(unique_labels(y_true, y_pred))`

jnothman commented on sklearn/metrics/classification.py:
```diff
@@ -1530,6 +1530,12 @@ class 2       1.00      0.67      0.80         3
 
+    # labelled micro average
+    label_micro_average = False
+    if y_type.startswith('multiclass') and \
```
I think `== 'multiclass'` should be more correct... but you also need to check for binary.

eamanu replied:
done

jnothman commented on sklearn/metrics/classification.py:
```diff
@@ -1530,6 +1530,12 @@ class 2       1.00      0.67      0.80         3
         labels = np.asarray(labels)
 
+    # labelled micro average
+    label_micro_average = False
```
It might be simpler to do this without `if`:

```
label_micro_average = (y_type == 'multiclass' and
                       (not labels_given or ...))
```

jnothman replied:
"label_micro_average" is a confusing name. Perhaps `micro_is_accuracy`?

eamanu replied:
done

amueller commented:
I think I would prefer either not showing it or showing it differently. Showing the same number 3 times is still confusing imho.

eamanu committed:
fix comments jnothman
In this commit change name of variable from label_micro_average to micro_is_accuracy
Add de label binary

eamanu commented:
Hi @amueller! ok, I'm going to remove it from the report. And then, it will great, wait for other developer opinion.

eamanu committed:
Remove micro average from report

eamanu committed:
Modify test_classification and add doc

amueller commented on sklearn/metrics/classification.py:
```diff
@@ -1502,6 +1502,11 @@ def classification_report(y_true, y_pred, labels=None, target_names=None,
         is also known as "sensitivity"; recall of the negative class is
         "specificity".
 
+        Show micro-average for multi-class classification is confusing,
```
I would change the above, put the micro average last in the list and say it's only shown for multi-label or multi-class with a subset of classes because it's accuracy otherwise.

eamanu replied:
done

eamanu committed:
fix doc

eamanu committed:
fix docs

jnothman reviewed (commented):
I think we should still be displaying accuracy. It is a very common metric for multiclass.

jnothman reviewed (commented):
Would the following be okay by you, @amueller?

```
                  precision    recall  f1-score   support
    <BLANKLINE>
         class 0       0.67      1.00      0.80         2
        class 1       0.00      0.00      0.00         1
        class 2       1.00      0.50      0.67         2
   <BLANKLINE>
      macro avg       0.56      0.50      0.49         5
   weighted avg       0.67      0.60      0.59         5
       accuracy                           0.60         5
```

and in the `dict_output=True`, do we just return a single value, i.e. `{..., 'accuracy': .3235, ...}`?

eamanu commented:
I will wait for @amueller response

eamanu commented:
@jnothman Do you consider necessary that accuracy appear at the final of the report? If it, I will have to modify other code apart of `classification_report()`

eamanu commented:
Anyway I have to modify the code of `format` to show just accuracy once

eamanu committed:
apply @jnothman comments

eamanu committed:
change on docs

jnothman reviewed (approved):
This looks good, IMO!

jnothman commented:
Except that there are test failures. @amueller, do you want this in 0.20.1?

eamanu commented:
@jnothman I think that I have to change the report (on dict_output=True) to `{..., 'accuracy': {'accuracy': .3235}, ...}` what do you think about it?

Because there are some tests that fail, e.g. `assert report[key].keys() == expected_report[key].keys()`

amueller commented:
I find it a bit weird that accuracy is in the f1-score column but I guess I'm fine with this. No strong opinion on 0.20.1 but I don't see a reason not to back-port if it gets merged in time.

amueller commented:
still test failures.

amueller commented:
Not sure if a blank line before accuracy would make it more readable? But I guess it's less confusing than what it was before...

jnothman commented:
You'll need to update the tests to match the dict output we want. But I'd
also like you to check, by hand, that pandas can convert the proposed
output into a reasonable DataFrame

eamanu commented:
@jnothman IMO there are not problems using pandas:

```
>>> a
{'macro avg': {'support': 75, 'precision': 0.5260083136726211, 'recall': 0.596146953405018, 'f1-score': 0.5099797365754813}, 'virginica': {'support': 20, 'f1-score': 0.5714285714285715, 'recall': 0.9, 'precision': 0.4186046511627907}, 'setosa': {'support': 24, 'f1-score': 0.8085106382978724, 'recall': 0.7916666666666666, 'precision': 0.8260869565217391}, 'versicolor': {'support': 31, 'f1-score': 0.15000000000000002, 'recall': 0.0967741935483871, 'precision': 0.3333333333333333}, 'weighted avg': {'support': 75, 'precision': 0.5137535108414785, 'recall': 0.5333333333333333, 'f1-score': 0.47310435663627154}, 'accuracy': 0.5333333333333333}
>>> pandas.DataFrame.from_dict(a)
           accuracy  macro avg     setosa  versicolor  virginica  weighted avg
f1-score   0.533333   0.509980   0.808511    0.150000   0.571429      0.473104
precision  0.533333   0.526008   0.826087    0.333333   0.418605      0.513754
recall     0.533333   0.596147   0.791667    0.096774   0.900000      0.533333
support    0.533333  75.000000  24.000000   31.000000  20.000000     75.000000

```

eamanu committed:
change on  test classification report dict

amueller commented:
For some reason I thought this was merged, hm...

amueller reviewed (approved):
Does this need a whatsnew?

eamanu commented:
@amueller I will create the whatsnew. hmm I think that I will try to merge this PR from the actual state. Maybe there  are some conflicts, aren't?

eamanu commented:
@amueller @jnothman I have ready this PR to merge. But, I wanted to (in local) merge with master, and I have (obviously) a lot of conflicts. Despite of this, I solved the conflicts but this results on 307 commits! + the whatsnew commit. What do you prefer? Create a new PR or made the push?

jnothman commented:
> But, I wanted to (in local) merge with master, and I have (obviously) a lot of conflicts

This is not obvious at all. What commands are you using to merge?

jnothman reviewed (commented):
Yes, this requires a what's new entry

jnothman commented:
Please add an entry to the change log at `doc/whats_new/v0.21.rst`. Like the other entries there, please reference this pull request with `:issue:` and credit yourself (and other contributors if applicable) with `:user:`

eamanu commented:
> This is not obvious at all. What commands are you using to merge?

From the branch fix_12334_micro-average_confusing i do `git merge mater`

jnothman commented:
Well git does not have a "Resolve Conflicts" button which means there are no conflicts with master. Perhaps your local master is not synced with master here.

eamanu commented:
> Well git does not have a "Resolve Conflicts" button which means there are no conflicts with master. Perhaps your local master is not synced with master here.

hmm, sorry, i don't tell you the whole sequence of commands. I run the next commands:

```
git checkout master
git pull origin master  # to sync my repo
git pull upstream master # to sync with scikit-learn
git push origin mater # to update my github
git checkout fix_12334_micro-average_confusing
git merge master
```

Then I have some conflicts. So, I resolve the conflicts manually.

So, what can I do? push my branch or with conflicts resolved by myself, or would you like see the conflicts?

jnothman commented:
```sh
git checkout master
git reset --hard upstream/master
git push --force origin master
git checkout fix_12334_micro-average_confusing
git merge master
```

eamanu commented:
> ```shell
> git checkout master
> git reset --hard upstream/master
> git push --force origin master
> git checkout fix_12334_micro-average_confusing
> git merge master
> ```

I run that commands and I still have conflicts. 

![image](https://user-images.githubusercontent.com/7605307/49839730-19cfce80-fd8f-11e8-9396-3b73fce9e8d8.png)


So, I decide create a New PR.

eamanu commented:
I create #12758. So I close this PR.
